weak_supervision/state_machines/transition_functions/basic_transition_function.py

```diff
@@ -0,0 +1,405 @@
+                    group_indices.extend([group_index] * len(actions))
+                    group_log_probs.append(log_probs)
+                    group_action_embeddings.append(action_embeddings)
+                    group_actions.extend(actions)
+                    group_curr_log_probs.append(curr_log_probs)
+
+
+                log_probs = torch.cat(group_log_probs, dim=0)
+                curr_log_probs = torch.cat(group_curr_log_probs, dim=0)
+                action_embeddings = torch.cat(group_action_embeddings, dim=0)
+                log_probs_cpu = log_probs.data.cpu().numpy().tolist()
+                batch_states = [(log_probs_cpu[i],
+                                 group_indices[i],
+                                 log_probs[i],
+                                 action_embeddings[i],
+                                 group_actions[i])
+                                for i in range(len(group_actions))
+                                if (not allowed_actions or
+                                    group_actions[i] in allowed_actions[group_indices[i]])]
+                # this is the only change required to sample from current log probabilities instead of beam search
+                if sample:
+                    categorical_dist = Categorical(logits=curr_log_probs)
+                    _, group_index, log_prob, action_embedding, action = batch_states[categorical_dist.sample()]
+                    new_states.append(make_state(group_index, action, log_prob, action_embedding))
+
+                else:
+                    batch_states.sort(key=lambda x: x[0], reverse=True)
+                    if max_actions:
+                        batch_states = batch_states[:max_actions]
+                    for _, group_index, log_prob, action_embedding, action in batch_states:
+                        new_states.append(make_state(group_index, action, log_prob, action_embedding))
+
+        return new_states
+
+    def _take_first_step(self,
+                         state: GrammarBasedState,
+                         allowed_actions: List[Set[int]] = None) -> List[GrammarBasedState]:
+        # We'll just do a projection from the current hidden state (which was initialized with the
+        # final encoder output) to the number of start actions that we have, normalize those
+        # logits, and use that as our score.  We end up duplicating some of the logic from
+        # `_compute_new_states` here, but we do things slightly differently, and it's easier to
+        # just copy the parts we need than to try to re-use that code.
+
+        # (group_size, hidden_dim)
+        hidden_state = torch.stack([rnn_state.hidden_state for rnn_state in state.rnn_state])
+        # (group_size, num_start_type)
+        start_action_logits = self._start_type_predictor(hidden_state)
+        log_probs = torch.nn.functional.log_softmax(start_action_logits, dim=-1)
+        sorted_log_probs, sorted_actions = log_probs.sort(dim=-1, descending=True)
```

Comment:
You'll have to make changes here too, to sample the first action from `log_probs` above.

Reply:
Oh i didn't see this. why is the first step handled in a different way than the rest of the steps?

Reply:
https://github.com/allenai/allennlp-weak-supervision-research/blob/b6b4293532be779de7dcc9c61ee517e3392a9ae5/weak_supervision/state_machines/transition_functions/basic_transition_function.py#L25-L29

Reply:
ok, so i'm going to add the same sampling logic to _take_first_step as well